### ISSUE TITLE
Consumers not getting messages after broker restart

### ISSUE THREAD

**Issue by EdeMeijer** [jhalterman/lyra#75]
Although this project sounds very convenient, I can't get it to work (at least with a consumer).

What I want to achieve is to start a consumer, then shut down the broker and restart it, and expecting the consumer to receive new messages.

I'm running RabbitMQ in Docker. Try this to reproduce the issue:

Run rabbitMQ in the foreground with
```bash
docker run -p 15672:15672 -p 5672:5672 --rm -it rabbitmq:3-management
```

Run this piece of java code (note I was trying to determine if there were types of exceptions I could add to the list of retryable/recoverable exceptions by using listeners, but nothing got logged).
```java
import com.rabbitmq.client.*;
import net.jodah.lyra.ConnectionOptions;
import net.jodah.lyra.Connections;
import net.jodah.lyra.config.Config;
import net.jodah.lyra.config.RecoveryPolicies;
import net.jodah.lyra.config.RetryPolicies;
import net.jodah.lyra.event.DefaultChannelListener;
import net.jodah.lyra.event.DefaultConnectionListener;
import net.jodah.lyra.event.DefaultConsumerListener;
import net.jodah.lyra.util.Duration;

import java.io.IOException;
import java.util.concurrent.TimeoutException;

public class App {
    public static void main(String[] args) throws TimeoutException, IOException {

        Config config = new Config()
            .withRecoveryPolicy(
                RecoveryPolicies.recoverAlways()
                    .withBackoff(Duration.seconds(1), Duration.seconds(30))
            )
            .withRetryPolicy(
                RetryPolicies.retryAlways()
                    .withBackoff(Duration.seconds(1), Duration.seconds(30))
            )
            .withChannelListeners(new DefaultChannelListener() {
                @Override
                public void onRecoveryFailure(Channel channel, Throwable failure) {
                    System.out.println("CHANNEL RECOVERY FAIL");
                    failure.printStackTrace();
                }
            })
            .withConnectionListeners(new DefaultConnectionListener() {
                @Override
                public void onRecoveryFailure(Connection connection, Throwable failure) {
                    System.out.println("CONNECTION RECOVERY FAIL");
                    failure.printStackTrace();
                }
            })
            .withConsumerListeners(new DefaultConsumerListener() {
                @Override
                public void onRecoveryFailure(Consumer consumer, Channel channel, Throwable failure) {
                    System.out.println("CONSUMER RECOVERY FAIL");
                    failure.printStackTrace();
                }
            });

        ConnectionOptions options = new ConnectionOptions().withHost("localhost");
        Connection connection = Connections.create(options, config);

        Channel channel = connection.createChannel();
        channel.queueDeclare("foo", true, false, false, null);
        channel.basicConsume("foo", new FooConsumer(channel));
    }

    static class FooConsumer extends DefaultConsumer {
        public FooConsumer(Channel channel) {
            super(channel);
        }

        @Override
        public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
            System.out.println(new String(body));
            getChannel().basicAck(envelope.getDeliveryTag(), false);
        }
    }
}
```
In your browser go to the management console on `localhost:15672`. Go to the `foo` queue and publish some messages. You will see them being printed by the java program.

Now terminate the broker by hitting ctrl + c in the terminal where you started the docker container.

Just restart it with up + enter.

In the output log from RabbitMQ, you can tell that the java app reconnects:
```
=INFO REPORT==== 23-Feb-2017::17:57:37 ===
accepting AMQP connection <0.1105.0> (172.17.0.1:42122 -> 172.17.0.2:5672)
```

Also, in the management UI you can see that a new consumer appears at the `foo` queue, so lyra is definitely reconnecting and starting a new consumer behind the scenes.

However, when publishing new messages through the UI, nothing gets printed on the console where the java program runs. In the UI, you can see that the messages are being consumed, but not acknowledged.

Maybe it's related to https://github.com/jhalterman/lyra/issues/74.

I hope you can find out what's going on or point out what I'm doing wrong or misunderstanding. I'd really like to use this in my project to save myself the hassle of building my own error handling/reconnecting stuff.

Thanks!

**Comment by EdeMeijer** [comment]
Hello,

Just wanted to ping you to check if you maybe have time to get back to me on this :)

Kind regards.

**Comment by michaelklishin** [comment]
@EdeMeijer I doubt this has anything to do with Lyra. You are not specifying how exactly you are publishing messages but it sounds like a routing problem and not a recovery one. See [Message Tracing](http://www.rabbitmq.com/firehose.html) and [how to capture traffic with Wireshark](https://www.rabbitmq.com/amqp-wireshark.html).

Also consider posting questions to rabbitmq-users and providing the entire server log.

**Comment by michaelklishin** [comment]
Also consider handling exceptions in your consumer's methods. While reasonably recent RabbitMQ Java client versions log unhandled exceptions and provide both forgiving and strict consumer implementations (the strict one closes channel upon an unhandled exception), it is much easier to inspect any possible exceptions if you handle them in your own code.

**Comment by neuhaus** [comment]
It's disappointing to see the recovery broken for lyra, after all that's the reason that people use it.
@jhalterman any input on this issue and #74?

**Comment by basert** [comment]
I was able to reproduce this issue by noticing that after the connection was restored by lyra, two consumers were on my queue. One was with this initial consumer tag, the other was with a new one.

I did some breakpoint debugging on the basicConsume method in `com.rabbitmq.client.impl.ChannelN` and the method gets called twice after the connection restore. This is due to changed behavior in rabbitmq > 4.0.0 where automatic recovery is enabled by default, resulting in a new consumer from lyra and the recovered by the rabbitmq library.

My solution was to disable this in the connection factory:
```
val connectionFactory = new com.rabbitmq.client.ConnectionFactory
connectionFactory.setAutomaticRecoveryEnabled(false)
val options = new ConnectionOptions(connectionFactory)
```

**Comment by jhalterman** [comment]
I'm no longer actively working on Lyra, but would be happy to take a PR to disable automatic recovery when using Lyra.

**Comment by basert** [comment]
added a PR, not sure how to release a new version though, I think that is up to you :)

**Comment by michaelklishin** [comment]
Addressed in #82.
